src/cassette.jl

```diff
@@ -141,6 +143,143 @@ function replacefunc(f::Function, d::AbstractDict)
+
+a structure to hold metadata for recursive type information for every function called
+            
+"""
+mutable struct Frame{F,T,U}
+    func::F
+    args::T
+    ret::U
+end
+            
+# add boilerplate for functionality
+function Cassette.overdub(ctx::TypeCtx, f, args...)
+    c = FCollector(ctx.metadata.depth-1, Frame(f, args, Any))
+    push!(ctx.metadata.data, c)
+    if c.depth > 0 && Cassette.canrecurse(ctx, f, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = c)
+        z = Cassette.recurse(newctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    else
+        z = Cassette.fallback(ctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    end
+end
+
+# limit the stacktrace in terms of which to recurse on
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Base.vect), args...) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(FCollector)) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Frame)) = false
+
+# this is a function that strips all using statements from a raw text file
+# reasons to do so
+# 1) we no longer include libraries we dont have too so semanticmodels can remain small
+# 2) prevents errors in the include-cassette design which doesn allow for usings
+function extractdeps(filename::String)
+    f = open(filename)
+    newf = open("new$filename","w") 
+    for line in readlines(f)
+        if occursin("using",line)
+            ex = Meta.parse(line)
+            eval(ex)
+        else
+            write(newf,"$line\n")
+        end
+    end
+    close(f)
+    close(newf)
+    return "new$filename"
+end
+                
+""" buildgraph
+
+internal function used in the typegraphfrompath
+takes the collector object and returns a metagraph
+            
+"""
+function buildgraph(g,collector)
+    add_vertex!(g,:name,collector.frame.args)
+    add_vertex!(g,:name,collector.frame.ret)
+    add_edge!(g,nv(g)-1,nv(g),:name,collector.frame.func)
+    for frame in collector.data
+        buildgraph(g,frame)
+    end
+    return g
+end
+
+""" typegraph(path::AbstractString,maxdepth::Int)
+            
+This is a function that takes in an array of script and produces a MetaDiGraph descibing the system.
+takes in optional parameter of recursion depth on the stacktrace defaulted to 3
+
+"""
+function typegraph(path::AbstractString,maxdepth::Int=3)
+    
+    # init the collector object
+    extractor = FCollector(maxdepth, Frame(nothing, (), nothing,))
+                
+    # init the context we want
+    ctx = TypeCtx(metadata = extractor);
+    
+    # build new script we want and load deps in env
+    newpath = extractdeps(path)
+                    
+    # cassette requires a callable stack
+    transcribe() = include(newpath)
+                        
+    # run the script internally and build the extractor data structure
+    Cassette.overdub(ctx,transcribe);
+                        
+    # delete our file we created for eval
+    rm(newpath)
```

Comment:
I don't like that this creates temporary files and removes them. If we can avoid touching the filesystem entirely, by either manipulating `Expr`s or `IOBuffer`. Exprs is my preferred option, but IO Buffers are better than creating files only to remove them.

If we want to treat the generated code as an artifact, then we should have a plan for how to store and organize it so that users can get the most out of it.

Reply:
If we change to 
```
f = open(file)
Cassette.overdub(TypeCtx,eval,readlines(f))
```
we can remove this feature.

Reply:
woah, that will trace through the eval call. what about:

```julia
f = open(file)
s = read(file)
ast = Base.Meta.parse(s)
m = eval(ast)
Cassette.overdub(TypeCtx, m.main, args...)
```

Evidence this will work:
```julia
julia> ast = Base.Meta.parse("module Moo main() = 1 end")
:(module Moo # none, line 1: # none, line 1:
    main() = begin  # none, line 1:
            1
        end
    end)

julia> m = eval(ast)
Moo

julia> m.main()
1
```

Reply:
This is weird but in my binary of julia the default return type of `read` is not str but `UInt8` 
![image](https://user-images.githubusercontent.com/41996797/53343456-6a5ebe80-38de-11e9-9e03-9ea0e8a6bc21.png)
If I change the default type this works
![image](https://user-images.githubusercontent.com/41996797/53343497-806c7f00-38de-11e9-8ed0-20a714d76b90.png)
How if we dont know the names of the funcs how do we hand this off to overdub? 
EG
![image](https://user-images.githubusercontent.com/41996797/53344515-d2ae9f80-38e0-11e9-8479-0d60ea385a83.png)
![image](https://user-images.githubusercontent.com/41996797/53344588-07baf200-38e1-11e9-9f92-fa00ab5f7f74.png)

Reply:
If this was the case, we might as well have mandated that all users do this instead...
```
function main()
    ...full script...
end
```

Reply:
Yeah the benefit of wrapping in a module is local variables won't leak and function names won't clash.